### DISCUSSION TITLE
How to handle missing data?

### DISCUSSION THREAD

**Discussion by csducate** [jspsych/jsPsych#1293] [Q&A] [answered]
Hi,

I've collected my first pilot data using jsPsych 6.1.0 and cognition.run, but I am running into trouble parsing the JSON strings (using jsonlite in R) when participants didn't answer a survey question. For instance, one line looks like this:

```
"{\"Q21.21\":4,\"Q21.22\":2,\"Q21.23\":2,\"Q21.24\":\"}"
``` 

I would like to change the default so that missing data is recorded as `NA` or `Null`, like this:
```
"{\"Q21.21\":3,\"Q21.22\":3,\"Q21.23\":3,\"Q21.24\":NA}"
``` 
Perhaps there is an easier way to handle this--this is my first effort at working with JSON--but right now, `jsonlite` balks at the first formulation, and all I can think to do is change how missing data is recorded.

Thanks in advance!

Caitlin

**Comment by csducate** [comment, marked as answer]
Oh, I found where to edit it in the jspsych-survey-multi-choice.js file!

```
      // create object to hold responses
      var question_data = {};
      for(var i=0; i<trial.questions.length; i++){
        var match = display_element.querySelector('#jspsych-survey-multi-choice-'+i);
        var id = "Q" + i;
        if(match.querySelector("input[type=radio]:checked") !== null){
          var val = match.querySelector("input[type=radio]:checked").value;
        } else {
          var val = null; // <--changed from one double quote (") to null
        }
```

**Reply by jodeleeuw** [reply]
This sounds like a bug on our end. Thanks for finding it! @becky-gilbert do you want to double check and add an issue if needed?

**Reply by becky-gilbert** [reply]
Thanks for reporting this @csducate! In [that line in v6.1.0 of the survey-multi-choice plugin](https://github.com/jspsych/jsPsych/blob/v6.1.0/plugins/jspsych-survey-multi-choice.js#L172), it looks like a missing response is saved as an empty string with two double quotes (not one):
```js
var val = "";
```
Also, when I test this with v6.1.0 and don't respond to one of the questions, this is the JSON string that I get in the data:
```
"{\"Q0\":\"Strongly Agree\",\"Q1\":\"\"}"
```
For Q1 in the string above, I get two quotes indicating an empty string, which differs from the way that the missing response is dealt with in the string from your first post. And I think this version would parse successfully with `jsonlite`, but I can double-check if that would be helpful?

So I'm not sure why your plugin had `var val = "`... Maybe it was accidentally modified at some point? Sorry not to be more helpful!

**Reply by csducate** [reply]
Thanks for getting back to me! I can't rule out that it might have gotten changed, but I've not edited a plugin until today, so I'm not sure how it would've happened. It would actually be helpful to see if `jsonlite` handles the double quote, though! I hadn't considered that the single double-quote vs. two double-quotes would be the problem.

Since I ran my study on cognition.run, I'll reach out to the creator, since the plugins are handled under the hood. Your response makes me think that the bug may ultimately be on their end.

**Reply by becky-gilbert** [reply]
I'm still not sure why you encountered this problem with the way that the missing responses were saved, and would really like to figure it out in case it's a bug on our end! Unfortunately I haven't been able to reproduce it... Here's what I've done so far: 

I ran the 6.1.0 survey-multi-choice example file, missed some questions ("Fruit" and "Q1"), and saved the data in JSON format. I'm getting the two double-quotes for non-responses. Here are my results from the txt file:
```json
[{"rt":2417.270000092685,"responses":"{\"Vegetables\":\"Strongly Disagree\",\"Fruit\":\"\"}","question_order":"[0,1]","trial_type":"survey-multi-choice","trial_index":0,"time_elapsed":2431,"internal_node_id":"0.0-0.0"},{"rt":1548.3099999837577,"responses":"{\"Q0\":\"Strongly Agree\",\"Q1\":\"\"}","question_order":"[0,1]","trial_type":"survey-multi-choice","trial_index":1,"time_elapsed":3991,"internal_node_id":"0.0-1.0"}]
```
I then read the results txt file into R and parsed them with `jsonlite`.
```r
library(jsonlite)
data_file <- file('C:\\Users\\bg04\\Documents\\test_results.txt', open="r")
raw_data <- readLines(data_file, warn=FALSE)
close(data_file)
parsed_results <- fromJSON(raw_data[1])
```
This is the value of `parsed_results`:
```r
       rt                                     responses question_order          trial_type trial_index time_elapsed internal_node_id
1 2417.27 {"Vegetables":"Strongly Disagree","Fruit":""}          [0,1] survey-multi-choice           0         2431          0.0-0.0
2 1548.31               {"Q0":"Strongly Agree","Q1":""}          [0,1] survey-multi-choice           1         3991          0.0-1.0
```
The `responses` column still contains JSON-formatted text, so I then parsed each of the `responses` cells separately:

```r
page1_responses <- as.data.frame(fromJSON(parsed_results$responses[1]))
page2_responses <- as.data.frame(fromJSON(parsed_results$responses[2]))
```
This parses successfully with `fromJSON` and gives me a data frame with columns for each question and a single row containing the responses. The Q1 (non-response) cell contain an empty string.
```
              Q0 Q1
Strongly Agree                   # <- empty string for Q1 (just appears blank in the R console)
```

> Since I ran my study on cognition.run, I'll reach out to the creator, since the plugins are handled under the hood. Your response makes me think that the bug may ultimately be on their end.

I also tried running the 6.1.0 survey-multi-choice example code on cognition.run, and I still get the double-quotes for missing responses. I added `localSave` to this demo, in case you want to try it here: https://powtrua6rs.cognition.run/ The results should be saved locally in CSV format (since cognition.run automatically saves data in CSV format, I figured this is how the data from your experiment was saved). 

Unfortunately I'm out of ideas for what might've caused this bug, but please let me know if there's more we can do to help figure it out.
